src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -86,6 +87,26 @@ internal ImmutableArray<string> GetDirectorySnapshot()
             return ImmutableArray<string>.Empty;
         }
 
+        public void BlockRestartOrShutdown()
+        {
+            lock (_cde)
+            {
+                if (_cde.CurrentCount == 0)
```

Comment:
Countdownevent public members are threadsafe, so you should be able to simplify this code to just a single line _cde.AddCount()? Seems to me that if CurrentCount is 0, we don't need to reset, but to just call AddCount?

Reply:
I think this is really because of the usage pattern here. Though the members are thread safe, checking the property and subsequently calling method isn't. 

I think we can simplify this with a count that is atomically incremented/decremented when the notifications are suspended and resumed. Then the notification handler can notify conditional on that count.

Reply:
Calling `Reset ` is required before using the `CountdownEvent`, otherwise we will get `ArgumentOutOfRangeException `- 
signalCount is less than or equal to 0. Also Reset is not threadsafe - https://docs.microsoft.com/en-us/dotnet/api/system.threading.countdownevent?view=netcore-3.1#thread-safety

Reply:
We shouldn't need to use a CountDownEvent here....